### PR TITLE
Add features

### PR DESCRIPTION
**Pull request by Or-Tal**
added features model support for base batch_solver and generator batch solver

### CODE REVIEW

File: denoiser/evaluate.py [hunk 1]
```diff
@@ -49,9 +48,9 @@ def evaluate(args, model, data_loader, epoch):
                 # If device is CPU, we do parallel evaluation in each CPU worker.
                 if args.device == 'cpu':
                     pendings.append(
-                        pool.submit(estimate_and_run_metrics, clean, model, noisy, args, filename))
+                        pool.submit(estimate_and_run_metrics, clean, model, noisy, args, filename, include_ft))
                 else:
-                    estimate = get_estimate(model, noisy)
+                    estimate = get_estimate(model, noisy)[0] if include_ft else get_estimate(model, noisy)
```
Comment (mosem) [inline denoiser/evaluate.py:53]: see above comment.

File: denoiser/models/dataclasses.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from dataclasses import dataclass
+
+@dataclass
+class FeaturesConfig:
+    include_ft: bool = False
+    feature_model: str = 'hubert'
+    state_dict_path: str = '/cs/labs/adiyoss/shared/pretrained_weights/hubert/hubert_base_ls960.pt'
+    features_factor: float = 0.01
```
Comment (mosem) [inline denoiser/models/dataclasses.py:8]: Nice idea for adding a dataclass :)

Add newline at the end of file.

File: denoiser/models/dummy_hubert.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+class DummyHubert(torch.nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        self.device = device
+
+    def extract_feats(self, x):
+        return torch.rand((x.shape[0], HUBERT_CH, HUBERT_TIME)).to(self.device)
+
+    def forward(self, x):
+        return x
```
Comment (mosem) [inline denoiser/models/dummy_hubert.py:16]: nice dummy :)
newline at end of file.

**Review by mosem — APPROVED** [review]
Overall looks good. Some of my comments need addressing before merging.

File: conf/experiment/demucs_1.yaml [hunk 4]
```diff
@@ -28,4 +28,10 @@ demucs:
...
   rescale: 0.1
-  scale_factor: ${experiment.scale_factor}
\ No newline at end of file
+  scale_factor: ${experiment.scale_factor}
+
+features_model:
+  include_ft: False
+  feature_model: 'hubert'
+  state_dict_path: '/cs/labs/adiyoss/shared/pretrained_weights/hubert/hubert_base_ls960.pt'
+  features_factor: 0.01
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply conf/experiment/demucs_1.yaml:37]: done

File: denoiser/batch_solvers/adversarial_bs.py [hunk 5]
```diff
@@ -27,12 +28,23 @@ def run(self, data, cross_valid=False):
...
+        else:
+            estimate = prediction
+            features_loss = 0
+
         discriminator_fake_detached = discriminator(estimate.detach())
         discriminator_real = discriminator(clean)
         discriminator_fake = discriminator(estimate)
 
-        loss_discriminator = self._get_discriminator_loss(discriminator_fake_detached, discriminator_real)
+        loss_discriminator = features_loss + self._get_discriminator_loss(discriminator_fake_detached, discriminator_real)
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/adversarial_bs.py:45]: You are right. it should not be a part of the discriminator loss, but a part of the generator loss. fixed it, thanks.

File: denoiser/batch_solvers/adversarial_bs.py [hunk 6]
```diff
@@ -27,12 +28,23 @@ def run(self, data, cross_valid=False):
         generator = self._models[GENERATOR_KEY]
         discriminator = self._models[DISCRIMINATOR_KEY]
 
-        estimate = generator(noisy)
+        prediction = generator(noisy)
+
+        # get features regularization loss if specified
+        if self.include_ft:
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/adversarial_bs.py:34]: You are right, fixed it.

File: denoiser/batch_solvers/batch_solver.py [hunk 7]
```diff
@@ -75,4 +86,13 @@ def get_generator_for_evaluation(self, best_states):
...
+        pass
+
+    def get_features_loss(self, estimated_embedded_dim, signal_to_extract_features_from):
+        if not self.include_ft:
+            return 0
+        with torch.no_grad():
+            y_ft = self.ft_model.extract_feats(signal_to_extract_features_from)
+            estimated_embedded_dim = F.interpolate(estimated_embedded_dim, y_ft.shape[-1]).permute(0, 2, 1)
+            estimated_embedded_dim = F.interpolate(estimated_embedded_dim, y_ft.shape[-2]).permute(0, 2, 1)
+            return F.l1_loss(y_ft, estimated_embedded_dim) * self.ft_factor
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/batch_solver.py:98]: Done

File: denoiser/batch_solvers/batch_solver.py [hunk 8]
```diff
@@ -75,4 +86,13 @@ def get_generator_for_evaluation(self, best_states):
         """
         loads the best state dict seen so far and returns a generator model ready for evaluation
         """
-        pass
\ No newline at end of file
+        pass
+
+    def get_features_loss(self, estimated_embedded_dim, signal_to_extract_features_from):
+        if not self.include_ft:
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/batch_solver.py:92]: this is where I force the loss to 0 in case that include features flag is off, removed it from adversarial

File: denoiser/batch_solvers/batch_solver.py [hunk 9]
```diff
@@ -75,4 +86,13 @@ def get_generator_for_evaluation(self, best_states):
         """
         loads the best state dict seen so far and returns a generator model ready for evaluation
         """
-        pass
\ No newline at end of file
+        pass
+
+    def get_features_loss(self, estimated_embedded_dim, signal_to_extract_features_from):
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/batch_solver.py:91]: yeah I agree, I thought you may find this name more estimated_embedded_dim more intuitive as it is the embedded_dim representation of the signal which I stretch to fit the predicted features dim. I've changed the name in adversarial_bs to be  estimated_embedded_dim, and y_ft in the function to be features, please take a look and tell me if this seems more intuitive now

File: denoiser/batch_solvers/batch_solver.py [hunk 10]
```diff
@@ -75,4 +86,13 @@ def get_generator_for_evaluation(self, best_states):
...
+        if not self.include_ft:
+            return 0
+        with torch.no_grad():
+            y_ft = self.ft_model.extract_feats(signal_to_extract_features_from)
+            estimated_embedded_dim = F.interpolate(estimated_embedded_dim, y_ft.shape[-1]).permute(0, 2, 1)
+            estimated_embedded_dim = F.interpolate(estimated_embedded_dim, y_ft.shape[-2]).permute(0, 2, 1)
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/batch_solver.py:97]: I've changed the names maybe now it is more readable.
What I do here is quite simple:
1. take the embedded_dim intermediate representation of the predicted signal
2. stretch it in both the time and channel domains ( I do this twice, each axis separately)
3. take an L1 loss between the stretched estimated embedded dim representation and the extracted features (from the clean signal)

File: denoiser/batch_solvers/batch_solver_factory.py [hunk 11]
```diff
@@ -14,19 +15,21 @@ class BatchSolverFactory:
 
     @staticmethod
     def get_bs(args):
+        ft_config = FeaturesConfig(**args.experiment.features_model) if hasattr(args.experiment, "features_model") else None
+
         if 'adversarial' in args.experiment and args.experiment.adversarial:
             if args.experiment.model == "demucs":
-                generator = Demucs(**args.experiment.demucs)
+                generator = Demucs(**args.experiment.demucs, include_ft_in_output=ft_config.include_ft if ft_config is not None else False)
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/batch_solvers/batch_solver_factory.py:22]: Yup I agree, fixed that.

File: denoiser/enhance.py [hunk 12]
```diff
@@ -102,10 +103,11 @@ def enhance(args, model, out_dir, data_loader):
                                 noisy_sigs, clean_sigs,
                                 raw_lengths_pairs,
                                 basenames, noisy_sr,
-                                args.experiment.sample_rate))
+                                args.experiment.sample_rate,
+                                include_ft))
             else:
                 # Forward
-                estimate = get_estimate(model, noisy_sigs)
+                estimate = get_estimate(model, noisy_sigs)[0] if include_ft else get_estimate(model, noisy_sigs)
```
[the remark this answers is not in the document]
Comment (Or-Tal) [reply denoiser/enhance.py:110]: Ok I agree, fixed that.

File: denoiser/evaluate.py (hunk 1 above)
Comment (Or-Tal) [reply denoiser/evaluate.py:53]: fixed

File: denoiser/models/dataclasses.py (hunk 2 above)
Comment (Or-Tal) [reply denoiser/models/dataclasses.py:8]: done

File: denoiser/models/dummy_hubert.py (hunk 3 above)
Comment (Or-Tal) [reply denoiser/models/dummy_hubert.py:16]: done
